src/Microsoft.AspNetCore.Cors/Infrastructure/CorsService.cs

```diff
@@ -82,24 +97,43 @@ public CorsResult EvaluatePolicy(HttpContext context, CorsPolicy policy)
         public virtual void EvaluateRequest(HttpContext context, CorsPolicy policy, CorsResult result)
         {
             var origin = context.Request.Headers[CorsConstants.Origin];
-            if (StringValues.IsNullOrEmpty(origin) || !policy.AllowAnyOrigin && !policy.Origins.Contains(origin))
+            if (StringValues.IsNullOrEmpty(origin))
             {
+                _logger?.RequestDoesNotHaveOriginHeader();
                 return;
             }
 
+            if (!policy.AllowAnyOrigin && !policy.Origins.Contains(origin))
+            {
+                _logger?.RequestHasOriginHeader();
+                _logger?.PolicyFailure($"Request origin {origin} does not have permission to access the resource.");
```

Comment:
Do not use format strings in logging. Logging has it's own feature for this that's better